File: Content/LeagueSandbox-Scripts/Champions/Global/SummonerHeal.cs
using System;
using GameServerCore.Domain.GameObjects;
using static LeagueSandbox.GameServer.API.ApiFunctionManager;
using LeagueSandbox.GameServer.Scripting.CSharp;
using GameServerCore.Domain.GameObjects.Spell;
using GameServerCore.Domain.GameObjects.Spell.Missile;
using System.Numerics;

namespace Spells
{
    public class SummonerHeal : ISpellScript
    {
        public ISpellScriptMetadata ScriptMetadata { get; private set; } = new SpellScriptMetadata()
        {
            // TODO
        };

        public void OnActivate(IObjAiBase owner, ISpell spell)
        {
        }

        public void OnDeactivate(IObjAiBase owner, ISpell spell)
        {
        }

        public void OnSpellPreCast(IObjAiBase owner, ISpell spell, IAttackableUnit target, Vector2 start, Vector2 end)
        {
            var units = GetChampionsInRange(owner.Position, 850, true);
            IChampion mostWoundedAlliedIChampion = null;
            float lowestHealthPercentage = 100;
            float maxHealth;
            foreach (var value in units)
            {
                if (value != owner && owner.Team == value.Team)
                {
                    var currentHealth = value.Stats.CurrentHealth;
                    maxHealth = value.Stats.HealthPoints.Total;
                    if (currentHealth * 100 / maxHealth < lowestHealthPercentage && owner != value)
                    {
                        lowestHealthPercentage = currentHealth * 100 / maxHealth;
                        mostWoundedAlliedIChampion = value;
                    }
                }
            }

            if (mostWoundedAlliedIChampion != null)
            {
                PerformHeal(owner, spell, mostWoundedAlliedIChampion);
            }

            PerformHeal(owner, spell, owner);
        }

        public void OnSpellCast(ISpell spell)
        {
        }

        public void OnSpellPostCast(ISpell spell)
        {
        }

        private void PerformHeal(IObjAiBase owner, ISpell spell, IAttackableUnit target)
        {
            float healthGain = 75 + (target.Stats.Level * 15);
            if (target.HasBuff("HealCheck"))
            {
                healthGain *= 0.5f;
            }
            var newHealth = target.Stats.CurrentHealth + healthGain;
            target.Stats.CurrentHealth = Math.Min(newHealth, target.Stats.HealthPoints.Total);
            AddBuff("HealSpeed", 1.0f, 1, spell, target, owner);
            AddBuff("HealCheck", 35.0f, 1, spell, target, owner);
            //AddParticleTarget(owner, "global_ss_heal_02.troy", target);
            AddParticleTarget(owner, "global_ss_heal_speedboost.troy", target);
        }

        public void OnUpdate(float diff)
        {
        }
    }
}

File: GameServer/Content/Data/LeagueSandbox-Default/Champions/Global/SummonerHeal.lua
function onFinishCasting()
	local owner = getOwner()
	local myTeam = owner:getTeam()
	local units = getChampionsInRange( owner, 850, true )
	local lowestHealthPercentage = 100
	local mostWoundedAlliedChampion = nil
	local lowestHealthPercentage = 100
	for i=0,units.Count-1 do
		value = units[i]
		if getOwner():getTeam() == value:getTeam() and i ~= 0 then
			local currentHealth = value:GetStats().CurrentHealth
			local maxHealth = value:GetStats().HealthPoints.Total
			if (currentHealth*100/maxHealth < lowestHealthPercentage) and (getOwner() ~= value) then
				lowestHealthPercentage = currentHealth*100/maxHealth
				mostWoundedAlliedChampion = value
			end
		end
	end
	
	if mostWoundedAlliedChampion then
		local newHealth = mostWoundedAlliedChampion:GetStats().CurrentHealth + 75 + getOwner():GetStats():GetLevel()*15
		local maxHealth = mostWoundedAlliedChampion:GetStats().HealthPoints.Total    
		if newHealth >= maxHealth then
			mostWoundedAlliedChampion:GetStats().CurrentHealth = maxHealth
		else
			mostWoundedAlliedChampion:GetStats().CurrentHealth = newHealth
		end
		
		local buff = Buff.new("Haste", 1.0, mostWoundedAlliedChampion, getOwner())
		buff:setMovementSpeedPercentModifier(30)
		addBuff(buff)
		
		addParticleTarget(mostWoundedAlliedChampion, "global_ss_heal_02.troy", mostWoundedAlliedChampion )
		addParticleTarget(mostWoundedAlliedChampion, "global_ss_heal_speedboost.troy", mostWoundedAlliedChampion )
	end
	newHealth = getOwner():GetStats().CurrentHealth + 75 + getOwner():GetStats():GetLevel()*15
	maxHealth = getOwner():GetStats().HealthPoints.Total
	if newHealth >= maxHealth then
		getOwner():GetStats().CurrentHealth = maxHealth
	else
		getOwner():GetStats().CurrentHealth = newHealth
	end
	
	addBuff("Haste", 1.0, owner, owner)
	
	addParticleTarget(getOwner(), "global_ss_heal.troy", getOwner())	
	addParticleTarget(getOwner(), "global_ss_heal_speedboost.troy", getOwner())
end

function applyEffects()
end
